### ISSUE TITLE
Configurable media quality

### ISSUE THREAD

**Issue by ara4n** [vector-im/element-call#249]
Currently our gUM constraints are hardcoded and very basic (640x360 video; no audio quality controls at all): https://github.com/matrix-org/matrix-js-sdk/blob/96ba061732bc3b5e748719072356d562e52097ba/src/webrtc/mediaHandler.ts#L296

We should let the user pick at least the video resolution & bitrate they want to send (presumably by some sensible defaults).

Also, leaving the bitrate entirely up to the browser feels risky in the extreme.

**Comment by w3bb** [comment]
Framerate is important for gamers, currently there isn't any alternatives to Discord that have framerates high enough to be worthwhile. An option for 30 and 60 FPS would be awesome.

**Comment by toniocat** [comment]
> Currently our gUM constraints are hardcoded and very basic (640x360 video; no audio quality controls at all): https://github.com/matrix-org/matrix-js-sdk/blob/96ba061732bc3b5e748719072356d562e52097ba/src/webrtc/mediaHandler.ts#L296
> 

If element-call is going to be available via Element Synapse server (web browser client) with users **not** installing their own Synapse client, hard-coding limits are preferred.  Perhaps we should have this ability enabled in only certain environments.

**Comment by terrytw** [comment]
I think it is sensible to at least make default resolution and bitrate configurable for homeservers.

**Comment by darkralts** [comment]
> Framerate is important for gamers, currently there isn't any alternatives to Discord that have framerates high enough to be worthwhile. An option for 30 and 60 FPS would be awesome.

i agree, 60fps is very important for me too, i need to watch and screen share movies at 60fps, as lots of media i have is 60fps vs 30, looks choppy on the others end.

**Comment by unknowntrojan** [comment]
The low quality for both audio and video is holding Element back from having more Discord users switch over. I'm not experienced with browser development, but I tried to tinker around with the code and see if I can force the browser to use a higher bitrate and achieve something better than 2kbps dishwater HD. So far I've not managed, but I imagine it can't be that complex, and I'm just missing something. Of course it's not only about the bitrate, being able to choose resolution & framerate is also very important. I'll keep trying, I guess.

**Comment by a11cf0** [comment]
Are there any plans to implement this?

**Comment by Moresteck** [comment]
> The low quality for both audio and video is holding Element back from having more Discord users switch over. I'm not experienced with browser development, but I tried to tinker around with the code and see if I can force the browser to use a higher bitrate and achieve something better than 2kbps dishwater HD. So far I've not managed, but I imagine it can't be that complex, and I'm just missing something. Of course it's not only about the bitrate, being able to choose resolution & framerate is also very important. I'll keep trying, I guess.

Hi, have you got anywhere with this?

**Comment by darkralts** [comment]
yeah like, discord is browser based electron, so it must be possible

**Comment by pirosuki** [comment]
This seems to work for screen share. No idea how you would make it easily configurable however.

Resolution and framerate
```diff   
diff --git a/src/room/InCallView.tsx b/src/room/InCallView.tsx
index d50be3c9..26515b3c 100644
--- a/src/room/InCallView.tsx
+++ b/src/room/InCallView.tsx
@@ -475,6 +475,11 @@ export const InCallView: FC<InCallViewProps> = ({
         selfBrowserSurface: "include",
         surfaceSwitching: "include",
         systemAudio: "include",
+        resolution: {
+          width: 1920,
+          height: 1080,
+          frameRate: 30,
+        },
       })
       .catch(logger.error);
```

Bitrate
```diff
diff --git a/src/livekit/options.ts b/src/livekit/options.ts
index f7fe03ff..c99ae98b 100644
--- a/src/livekit/options.ts
+++ b/src/livekit/options.ts
@@ -24,7 +24,10 @@ const defaultLiveKitPublishOptions: TrackPublishDefaults = {  
   forceStereo: false,
   simulcast: true,
   videoSimulcastLayers: [VideoPresets.h180, VideoPresets.h360] as VideoPreset[],
-  screenShareEncoding: ScreenSharePresets.h1080fps30.encoding,
+  screenShareEncoding: {
+    maxBitrate: 5_000_000,
+    maxFramerate: 30,
+  },
   stopMicTrackOnMute: false,
   videoCodec: "vp8",
   videoEncoding: VideoPresets.h720.encoding,
```

**Comment by Destinyg133** [comment]
Any news on this? ❤️

**Comment by 0x4c756e61** [comment]
Right now the best way me and a friend found to bypass those limitations, is to record the screen with obs and use the virtual camera it creates. The bandwidth is still extremely low, but it seems to be way better than what the screen sharing one is.

**Comment by DanceMore** [comment]
whenever I tell my friends about `element-call`, the is "does it support lossless audio? or at least mp3 listening quality??"

many of them are musicians and they understand they're not going to get low latency for performance or live collaboration, but a major usecase for screensharing and/or audio sharing is "hey check out this recording / project in my DAW!" people are willing to perform strange configuration or route audio in hardware, but if the protocol doesn't support the bitrate / packets, even line-in audio will sound bad on the other side.

I already tried using plain old Linux trickery to route audio playback to a pulseaudio virtual microphone and send it straight into `element-call` and listen on the other side with nice headphones.... it felt like [telephone hold music ](https://www.youtube.com/watch?v=N7xn5zeJ4D4)quality :( 

---

it sounds like in theory, livekit servers and most (all?) browsers can run opus codec up to ~514kbps and allegedly get listening quality audio across the wire ...? all the docs I found made it seem like it's not something I can toggle on in `livekit.yml` server-side, it's something the client actually needs to request similar to this patch https://github.com/element-hq/element-call/issues/249#issuecomment-2436601201

**Comment by whitehatrabb1t** [comment]
> This seems to work for screen share. No idea how you would make it easily configurable however.
> 
> 
> 
> Resolution and framerate
> 
> ```diff   
> 
> diff --git a/src/room/InCallView.tsx b/src/room/InCallView.tsx
> 
> index d50be3c9..26515b3c 100644
> 
> --- a/src/room/InCallView.tsx
> 
> +++ b/src/room/InCallView.tsx
> 
> @@ -475,6 +475,11 @@ export const InCallView: FC<InCallViewProps> = ({
> 
>          selfBrowserSurface: "include",
> 
>          surfaceSwitching: "include",
> 
>          systemAudio: "include",
> 
> +        resolution: {
> 
> +          width: 1920,
> 
> +          height: 1080,
> 
> +          frameRate: 30,
> 
> +        },
> 
>        })
> 
>        .catch(logger.error);
> 
> ```
> 
> 
> 
> Bitrate
> 
> ```diff
> 
> diff --git a/src/livekit/options.ts b/src/livekit/options.ts
> 
> index f7fe03ff..c99ae98b 100644
> 
> --- a/src/livekit/options.ts
> 
> +++ b/src/livekit/options.ts
> 
> @@ -24,7 +24,10 @@ const defaultLiveKitPublishOptions: TrackPublishDefaults = {  
> 
>    forceStereo: false,
> 
>    simulcast: true,
> 
>    videoSimulcastLayers: [VideoPresets.h180, VideoPresets.h360] as VideoPreset[],
> 
> -  screenShareEncoding: ScreenSharePresets.h1080fps30.encoding,
> 
> +  screenShareEncoding: {
> 
> +    maxBitrate: 5_000_000,
> 
> +    maxFramerate: 30,
> 
> +  },
> 
>    stopMicTrackOnMute: false,
> 
>    videoCodec: "vp8",
> 
>    videoEncoding: VideoPresets.h720.encoding,
> 
> ```

So is that not kinda adoptable for Calls in General?

I also feel the same need for good quality, we got 2025 and it looks like ..., can be way better..

**Comment by orgmast5** [comment]
2025 and my webcam still streams at 360p

**Comment by darkralts** [comment]
xd

**Comment by darkralts** [comment]
mines higher than that

**Comment by jameskitt616** [comment]
Any news on this 2022 issue? We are in 2026 now and Element-Call still wants us to stream mushy Video with low resolution, bitrate and framerate.
Please make it configureable if this should be any useable 🙏🏻

**Comment by orgmast5** [comment]
@fkwp is the 360p hardcoded in livekit? can we add a slider or enumeration on supported res by camera/screen?

**Comment by pirosuki** [comment]
Hi again!
I made a proof of concept for adding these options to the gui. I'll put it up on here in the evening so you all can have a look. Couldn't figure out how to put the video here in the issue so here's the link to a video of me using it to change resolution, framerate, bitrate and codec.

https://piro.sh/advancedscreensharesettings.mkv

edit:
mp4 for the mobile users

https://piro.sh/advancedscreensharesettings.mp4

**Comment by alexander-potemkin** [comment]
@pirosuki , apologies for the offtop, but how may I have my XFCE look the same? :)

**Comment by pirosuki** [comment]
@alexander-potemkin it's kde actually ^^

**Comment by pirosuki** [comment]
Here's the code y'all https://github.com/pirosuki/element-call/tree/advanced-screen-share-settings

**Comment by jameskitt616** [comment]
Awesome work, thank you!

Making defaults configurable via e.g. env or config file would also be of much benefit.

**Comment by xangelix** [comment]
This would be fantastic-- thank you so much!

Is a similar approach possible for normal video/audio? Non-screensharing?

Would be great to have raw numbers displayed for the sliders too.

**Comment by coolsrett3333** [comment]
> Here's the code y'all https://github.com/pirosuki/element-call/tree/advanced-screen-share-settings

this looks pretty good, could you rebase and send a PR to this repo for the maintainers to review the code?

**Comment by tylo** [comment]
> Here's the code y'all https://github.com/pirosuki/element-call/tree/advanced-screen-share-settings

Please, please, please @pirosuki make this a PR to element-call. The default settings are pretty poor reliability if you want a smooth framerate desktop stream.

**Comment by pirosuki** [comment]
o/

@jameskitt616 I'll look into that. I was thinking there should probably be a way to cap the bitrate to limit network load so maybe that too.

@xangelix I'm pretty sure livekit has support for auto gain, echo cancellation and noise suppression so I could probably add those as audio options. Adding the quality options to the normal video sharing shouldn't be an issue either. As for displaying raw numbers, while we absolutely could do that I don't think they would be very useful. The bitrate slider currently steps in 0.5 Mbps intervals you'd just see the same value as in the video but with a lot more zeroes behind it.

@coolsrett3333 @tylo As soon as I've looked into the things above I'll see if I can figure out how to make a pr ^^

**Comment by tylo** [comment]
I see these commits were added as references, but I feel like it does a different thing. Those commits allow the server owner to set new defaults, but don't allow the user to set their own settings.

I'd like both to exist, it would be nice if a server owner can set what default values people stream at, but it should ultimately be up to each individual user to be able to tweak their settings in real-time to get the best quality they can for their given situation.

**Comment by entropylaw** [comment]
how about now?

**Comment by tylo** [comment]
Looks like there is a PR up for a feature like this: https://github.com/element-hq/element-call/pull/3736

**Comment by GGort** [comment]
> Looks like there is a PR up for a feature like this: https://github.com/element-hq/element-call/pull/3736

That looks to be for 1:1 calling not element-call calls?

**Comment by tylo** [comment]
> > Looks like there is a PR up for a feature like this: [#3736](https://github.com/element-hq/element-call/pull/3736)
> 
> That looks to be for 1:1 calling not element-call calls?

No, I am pretty sure these settings exist in any element call, group or 1:1. What I shared is not what pirosuki mentioned on Feb 25th.

**Comment by davidkirch** [comment]
Hey everybody, thanks for the great prs so far! Is there any update on this? The lack of higher quality screen share is still a major reasons for not switching for us. Thanks!

**Comment by tylo** [comment]
Keep an eye on the PR I linked to. I don't know what is holding it up exactly, but open source works at the pace of "whenever the developer feels like it". PRs are a big bottleneck for that I'm afraid.
